flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/windowing/AsyncWindowOperator.java

```diff
@@ -0,0 +1,810 @@
+
+        if (windowAssigner instanceof MergingWindowAssigner) {
+            throw new UnsupportedOperationException(
+                    "Async WindowOperator not support merging window (e.g. session window) yet.");
+        } else {
+            for (W window : elementWindows) {
+
+                // drop if the window is already late
+                if (isWindowLate(window)) {
+                    continue;
+                }
+
+                AtomicReference<TriggerResult> triggerResult = new AtomicReference<>();
+                windowDeclaredVariable.set(window);
+
+                windowFutures.add(
+                        windowState
+                                .asyncAdd(element.getValue())
+                                .thenCompose(
+                                        ignore ->
+                                                triggerContext
+                                                        .onElement(element)
+                                                        .thenAccept(triggerResult::set))
+                                .thenConditionallyCompose(
+                                        ignore -> triggerResult.get().isFire(),
+                                        ignore ->
+                                                windowState
+                                                        .asyncGet()
+                                                        .thenConditionallyCompose(
+                                                                Objects::nonNull,
+                                                                contents ->
+                                                                        emitWindowContents(
+                                                                                key, window,
+                                                                                contents),
+                                                                StateFutureUtils
+                                                                        .createVoidReturnFunction()),
+                                        StateFutureUtils.createVoidReturnFunction())
+                                .thenConditionallyCompose(
+                                        ignore -> triggerResult.get().isPurge(),
+                                        ignore -> windowState.asyncClear(),
+                                        StateFutureUtils.createVoidReturnFunction())
+                                .thenApply(
+                                        empty -> {
+                                            registerCleanupTimer(window);
+                                            // if element is handled by none of assigned
+                                            // elementWindows
+                                            return false; // isSkippedElement = false
+                                        }));
```

Comment:
```suggestion
                                .thenAccept(
                                        empty -> {
                                            registerCleanupTimer(window);
                                        }));
```

Reply:
Thank you for your reminding. I have use `windowFutures` to implicitly indicate whether `isSkippedElement == true`, i.e., `windowFutures is empty => implicitly isSkippedElement == true`